### DISCUSSION TITLE
My study notes for ELSA 3

### DISCUSSION THREAD

**Discussion by dodyg** [elsa-workflows/elsa-core#3696] [General]
I study by creating samples. This one was mine back in [2021 on ELSA 2](https://github.com/elsa-workflows/elsa-core/discussions/760).

This is my series of samples so far on [ELSA 3](https://github.com/dodyg/practical-aspnetcore/tree/net6.0/projects/elsa). I am starting from the basic control and hopefully I will understand more enough to create more complex samples.

**Comment by dodyg** [comment]
Sample for composite activity

https://github.com/dodyg/practical-aspnetcore/tree/net6.0/projects/elsa/composite-activity

**Reply by theafricaninternetgroup** [reply]
I am following you closely.Great work @dodyg

**Comment by dodyg** [comment]
I got 12 samples so far. Still digging https://github.com/dodyg/practical-aspnetcore/blob/net6.0/projects/elsa/README.md

**Reply by sfmskywalker** [reply]
Amazing! We should link to your repo from the documentation. If you have some time, please feel free to open a PR if you want.

**Reply by sfmskywalker** [reply]
I added a link to your repo here: https://github.com/elsa-workflows/elsa-documentation/issues/31 so we can collect more resources created by community members such as yourself.

**Comment by dodyg** [comment]
`3.0.0-preview.18`

@sfmskywalker  this looks weird

```csharp

using Elsa.Extensions;
using Elsa.Workflows.Core.Activities;
using Elsa.Workflows.Core.Models;
using Elsa.Workflows.Core.Services;

var services = new ServiceCollection();
services.AddElsa();

var serviceProvider = services.BuildServiceProvider();
var runner = serviceProvider.GetRequiredService<IWorkflowRunner>();

await runner.RunAsync(new ConstructorWorkflow("Anne", 37));

public class ConstructorWorkflow : WorkflowBase
{
    private readonly Variable<string> _name;
    private readonly Variable<int> _age;

    public ConstructorWorkflow(string name, int age)
    {
        _name = new Variable<string>(name);
        _age = new Variable<int>(age);
    }

    protected override void Build(IWorkflowBuilder builder)
    {
        builder.Root = new Sequence
        {
            Variables = { _name, _age },
            Activities = 
            {
                new WriteLine(ctx => $"Name: {_name.Get(ctx)}"),
                new WriteLine(ctx => $"Age: {_age.Get(ctx)}")
            }
        };
    }
}
```

returns this result 
```
PS D:\GitHub\practical-aspnetcore\projects\elsa\workflow-5> dotnet run
D:\GitHub\practical-aspnetcore\projects\elsa\workflow-5\Program.cs(9,23): warning ASP0000: Calling 'BuildServiceProvider' from application code results in an additional copy of singleton services being created. Consi
der alternatives such as dependency injecting services as parameters to 'Configure'. [D:\GitHub\practical-aspnetcore\projects\elsa\workflow-5\workflow-5.csproj]
Name: 
Age: 37
```

**Reply by sfmskywalker** [reply]
Indeed. And it makes sense in the context of a program that uses `var builder = WebApplication.CreateBuilder(args);`, where the service container is built when doing `app.Build()`. Here, you don't want to (accidentally) call `BuildServiceProvider`.

But in a plain console program as this one, how else is one supposed to construct a service container? 🤔 
I may be wrong. but it seems to me that the warning is missing some context.

PS. Kudos on the nice sample! 😄

**Reply by dodyg** [reply]
`new WriteLine(ctx => $"Name: {_name.Get(ctx)}"),` prints `Name: ` although it is assigned at constructor.

**Reply by dodyg** [reply]
@sfmskywalker  it looks like a bug

**Comment by dodyg** [comment]
Right now I am trying to figure out where to get the values that were set on `WorkflowExecutionContext` at https://github.com/elsa-workflows/elsa-core/blob/v3/src/modules/Elsa.Workflows.Core/Activities/SetName.cs

Below's code for example doesn't work. There is no value in `   var instanceName = context.WorkflowExecutionContext.GetProperty<string>("WorkflowInstanceName");`.

``` csharp

public class SetNameWorkflow : WorkflowBase
{
    protected override void Build(IWorkflowBuilder builder)
    {
        builder.Root = new Sequence
        {
            Activities = 
            {
                new SetName("My Simple Workflow"),
                new ShowName()
            }
        };
    } 
}

public class ShowName : Activity
{
    protected override async ValueTask ExecuteAsync(ActivityExecutionContext context)
    {
        var instanceName = context.WorkflowExecutionContext.GetProperty<string>("WorkflowInstanceName");
        Console.WriteLine($"WorkflowInstanceName {instanceName}");
    }
}

```

**Comment by dodyg** [comment]
This doesn't work. There is no prompt at the console.

``` csharp
using Elsa.Extensions;
using Elsa.Workflows.Core.Activities;
using Elsa.Workflows.Core.Models;
using Elsa.Workflows.Core.Services;

var services = new ServiceCollection();
services.AddElsa();

var input = new Input<string>("What is your name");
var name = new Variable<string>("name");

var serviceProvider = services.BuildServiceProvider();
var workflow = new Sequence 
{
    Variables = 
    {
        name
    },
    Activities = 
    {
        new WriteLine(context => input.Get(context)),
        new ReadLine(name),
        new WriteLine(ctx => "My name is " + name.Get(ctx))
    }    
};

var runner = serviceProvider.GetRequiredService<IWorkflowRunner>();
await runner.RunAsync(workflow);
```

The whole sample is here https://github.com/dodyg/practical-aspnetcore/tree/net6.0/projects/elsa/readline-activity

**Comment by dodyg** [comment]
A simple `Fork Activity` sample added https://github.com/dodyg/practical-aspnetcore/tree/net6.0/projects/elsa/fork-activity.

Next I want to demonstrate the difference behaviour in `JoinMode`.

**Comment by dodyg** [comment]
I updated all the existing samples to the latest revision of Elsa 3

https://github.com/dodyg/practical-aspnetcore/tree/net6.0/projects/elsa

**Reply by sfmskywalker** [reply]
Thanks for doing this! For discoverability of your work, I think we should add a link from the documentation website and the README of the elsa-core repo. Feel free to open PRs to that effect if you want to.

**Reply by sfmskywalker** [reply]
I added a link to your repo here: https://github.com/elsa-workflows/elsa-documentation/issues/31 so we can collect more resources created by community members such as yourself.
